yt_dlp/postprocessor/modify_chapters.py

```diff
@@ -223,43 +251,59 @@ def append_chapter(c):
+                    after_c = dict(copy.deepcopy(cur_chapter), start_time=c['end_time'])
                     heapq.heappush(chapters, (after_c['start_time'], cur_i, after_c))
-                # Sponsor chapter must inherit the cuts that the normal chapter
-                # has accumulated within it.
+                # Push the part after the overlap to PQ.
+                elif c['end_time'] > cur_chapter['end_time']:
+                    after_cur = dict(copy.deepcopy(c), start_time=cur_chapter['end_time'])
+                    heapq.heappush(chapters, (after_cur['start_time'], cur_i, after_cur))
+                    c['end_time'] = cur_chapter['end_time']
+                # (sponsor, sponsor): merge categories in the overlap.
+                if '_categories' in cur_chapter:
+                    c['_categories'] = cur_chapter['_categories'] + c['_categories']
+                # Inherit the cuts that the current chapter has accumulated within it.
                 if 'cut_idx' in cur_chapter:
                     c['cut_idx'] = cur_chapter['cut_idx']
                 cur_chapter['end_time'] = c['start_time']
                 append_chapter(cur_chapter)
                 cur_i, cur_chapter = i, c
         (append_chapter if 'remove' not in cur_chapter else append_cut)(cur_chapter)
 
-        for c in new_chapters:
-            if 'categories' in c:
-                cats = sponsor_categories(c)
-                # Overlapping sponsor chapters will have a title that looks like
-                # '[SponsorBlock]: Sponsor/Interaction Remainder/Self-Promotion',
-                # in the order in which the categories were encountered.
-                # OrderedDict removes duplicates preserving the order.
-                c['title'] = '[SponsorBlock]: ' + '/'.join(OrderedDict(
-                    zip(cats, map(SponsorBlockPP.CATEGORIES.__getitem__, cats))).values())
-                del c['categories']
+        i = -1
+        for c in new_chapters.copy():
+            i += 1
+            c.pop('_was_cut', None)
+            cats = c.pop('_categories', None)
+            if cats:
+                category = min(cats, key=lambda c: c[2] - c[1])[0]
+                cats = orderedSet(x[0] for x in cats)
+                c.update({
+                    'category': category,
+                    'categories': cats,
+                    'name': SponsorBlockPP.CATEGORIES[category],
+                    'category_names': [SponsorBlockPP.CATEGORIES[c] for c in cats]
+                })
+                outtmpl, tmpl_dict = self._downloader.prepare_outtmpl(self._sponsorblock_chapter_title, c)
+                c['title'] = self._downloader.escape_outtmpl(outtmpl) % tmpl_dict
+            if i > 0 and c['title'] == new_chapters[i - 1]['title']:
+                new_chapters[i - 1]['end_time'] = c['end_time']
+                new_chapters.pop(i)
+                i -= 1
```

Comment:
This assumes that `new_chapters` is already sorted. This is always correct, no?

Reply:
Yes.